YouTube + Music

Titiftw opened the pull request:
Hi,
I found your project and I really enjoyed it. However, it was lacking few functionalities that I tried to add myself. Note that I am not a Python nor a Svelte expert but if it may help you, it includes Music upload and YouTube (that one might require some fanciness on the link input).
I'm missing software for the translations though.

If you think this can help you, feel free to use it. If not, then please discard that PR!

Kind regards,
Jason

mawoka-myblock commented on classquiz/db/models.py:
```diff
@@ -136,6 +136,8 @@ class QuizQuestion(BaseModel):
     type: None | QuizQuestionType = QuizQuestionType.ABCD
     answers: list[ABCDQuizAnswer] | RangeQuizAnswer | list[TextQuizAnswer] | list[VotingQuizAnswer] | str
     image: str | None = None
+    youtubeUrl: str | None = None
```
Use `snake_case` please

mawoka-myblock commented on classquiz/helpers/__init__.py:
```diff
@@ -297,3 +297,12 @@ def extract_image_ids_from_quiz(quiz: Quiz) -> list[str | uuid.UUID]:
+def extract_music_ids_from_quiz(quiz: Quiz) -> list[str | uuid.UUID]:
+    quiz_musics = []
+    for question in quiz.questions:
+        if question["music"] is None:
```
Use `.get()` to prevent KeyError

mawoka-myblock commented on classquiz/routers/editor.py:
```diff
@@ -127,12 +131,6 @@ async def finish_edit(edit_id: str, quiz_input: QuizInput):
         quiz.background_color = quiz_input.background_color
         quiz.background_image = quiz_input.background_image
         quiz.mod_rating = None
-        for image in images_to_delete:
-            if image is not None:
-                try:
-                    await storage.delete([image])
-                except DeletionFailedError:
-                    pass
```
Why remove that?

Titiftw replied:
I don't see "images_to_delete" being populated. Am I wrong? It seems it's always set to []

mawoka-myblock replied:
It *should* find the images that were in the quiz before it was saved and deletes them, but I'll check that

mawoka-myblock commented on classquiz/worker/storage.py:
```diff
@@ -96,7 +76,21 @@ async def quiz_update(ctx, old_quiz: Quiz, quiz_id: uuid.UUID):
+                continue
+            change_made = True
+    for music in removed_musics:
+        if "--" in music:
```
Not needed, as this ensures legacy compatibility

Titiftw replied:
Which part? The "--" check ?

mawoka-myblock replied:
Exactly. I left the comment for that line, but GitHub doesn't tell you so. So yes, the part with the `--`.

mawoka-myblock commented on frontend/package-lock.json:
Please remove that

mawoka-myblock commented on frontend/src/lib/editor/card.svelte:
```diff
@@ -152,6 +152,72 @@ SPDX-License-Identifier: MPL-2.0
 							<MediaComponent bind:src={image_url} />
 						</div>
 					</div>
+				{:else if data.questions[selected_question].youtubeUrl}
```
Snake case pls

mawoka-myblock commented on frontend/src/lib/editor/card.svelte:
```diff
@@ -152,6 +152,72 @@ SPDX-License-Identifier: MPL-2.0
+								height="328"
+								src={data.questions[selected_question].youtubeUrl}
+								frameborder="0"
+								allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share"
```
Are these permissions really needed?

mawoka-myblock commented on frontend/src/lib/editor/uploader/Library.svelte:
```diff
@@ -17,48 +17,159 @@ SPDX-License-Identifier: MPL-2.0
+			data.background_image = file.id;
 		} else {
-			data.questions[selected_question].image = id;
+			if (file.mime_type === 'music/mp3') {
```
There's way more than mp3, check for `audio/*` or `music/*`

mawoka-myblock commented on frontend/src/lib/editor/uploader/Library.svelte:
```diff
@@ -17,48 +17,159 @@ SPDX-License-Identifier: MPL-2.0
-							class="object-contain w-full h-full max-h-full rounded"
-						/>
+			{#each files as file}
+				{#if file.mime_type === 'music/mp3'}
```
Same mime-type oversight here

mawoka-myblock commented on frontend/src/lib/editor/uploader/Library.svelte:
```diff
@@ -17,48 +17,159 @@ SPDX-License-Identifier: MPL-2.0
+				{#if file.mime_type === 'music/mp3'}
+					<div class="rounded border-2 border-[#B07156] p-2 flex-col flex gap-2">
+						<div>
+							<svg
```
Any kind of source for this SVG?

mawoka-myblock commented on frontend/src/lib/quiz_types.ts:
```diff
@@ -60,6 +60,8 @@ export interface Question {
 	question: string;
 	type?: QuizQuestionType;
 	image?: string;
+	youtubeUrl?: string;
```
snake_case

mawoka-myblock commented on frontend/src/routes/edit/files/+page.svelte:
```diff
@@ -17,7 +17,7 @@ SPDX-License-Identifier: MPL-2.0
 	export let data: PageData;
 	let edit_popup = null;
-	const images = data.images;
+	const files = data.images;
```
Still `data.images`? Just asking

mawoka-myblock commented on frontend/src/routes/edit/files/+page.svelte:
```diff
@@ -59,57 +59,157 @@ SPDX-License-Identifier: MPL-2.0
-					loading="lazy"
-					alt={image.alt_text || $t('file_dashboard.not_available')}
-				/>
+				{#if file.mime_type === 'music/mp3'}
```
mime type oversight

Titiftw replied:
What do you mean ?

mawoka-myblock commented on frontend/src/routes/edit/files/+page.svelte:
```diff
@@ -59,57 +59,157 @@ SPDX-License-Identifier: MPL-2.0
-					alt={image.alt_text || $t('file_dashboard.not_available')}
-				/>
+				{#if file.mime_type === 'music/mp3'}
+					<svg
```
svg source?

Titiftw replied:
https://www.svgrepo.com/svg/2225/music

mawoka-myblock commented on frontend/src/routes/edit/musics/+page.svelte:
```diff
@@ -0,0 +1,151 @@
+			time_elapsed: 0
+		};
+
+	const speed_extraction_regex = /speed=(\d+\.\d+)x(?![\s\S]*speed=\d+\.\d+x)/;
```
Stale

mawoka-myblock commented on frontend/src/routes/edit/musics/+page.svelte:
```diff
@@ -0,0 +1,151 @@
+				});
+			});
+			xhr.open('POST', `/api/v1/storage/raw/${file.name}`, true);
+			xhr.setRequestHeader('Content-Type', 'music/mp3');
```
Not always mp3

Titiftw replied:
You're right but I don't have any other option (checking file extension?)

mawoka-myblock commented on frontend/src/routes/edit/musics/+page.svelte:
```diff
@@ -0,0 +1,151 @@
+	let file: File;
+	let file_data: undefined | Blob = undefined;
+
+	const compress_music = async () => {
```
What's this purpose?

Titiftw replied:
None, but I didn't manage and spend time to remove the way it used to work. So I kept the compression function if compression could be added.

mawoka-myblock commented:
Thank you soo much for this PR! Really appreciate it! I've just looked through the code roughly and put some notes down. Please address them or comment on them and mark them as `resolved` afterwards. Another thing I haven't seen: Where's the YouTube player? maybe I just missed it. I'll also have to test this code locally. Again, many thanks for such a huge PR!

Titiftw commented:
> Owner

The YouTube player is done using iframe:
![CleanShot 2024-09-25 at 17 35 47@2x](https://github.com/user-attachments/assets/056850d3-19ed-41a0-ae51-1f589d2741b1)
Once the link is added to the card, it gets displayed on it:
![CleanShot 2024-09-25 at 17 37 29@2x](https://github.com/user-attachments/assets/6b312f1b-3a0e-4034-b3f9-e4f0910159e5)
Might still be missing from the Admin card as I didn't test it fully as I realized that the YouTube video title would be showed to every body... Not the greatest idea for Blind tests.

Titiftw commented:
@mawoka-myblock I'm having issue "unlinking" images/musics from a quiz. If I remove a music or an image from a quiz and it becomes unused in any quiz, the call "/storage/list" still shows it as part of a quiz which prevents the file delete button to appear. It seems that the storage worker line `await new_quiz.storageitems.remove(item)` always fails as the item UUID doesn't exist.
Any idea what's wrong?

mawoka-myblock commented:
I'm gonna have to look into that when I got time. Expect the weekend in 3 days

Titiftw commented:
No worries @mawoka-myblock, take your time! In the meantime, I also worked on another feature to fix several points which were blocking my tests:

- [ ] When playing, there was a bug when a player would reconnect more than once which could potentially happen. I fixed it in my feature branch (now the player can refresh any time the game, as long as it is not answering a question).
- [ ] When playing, the total score showed on the player screen was wrong if the player reconnected.
- [ ] When playing, a player could give X answers if it would click X times on the "Validate" button. I adjusted the disabled button logic in the frontend and also return an "already_replied" on the WebSocket.
- [ ] When playing on a CHECK question, the player could select multiple answers. I made sure all the selected answers were removed when clicking on another answer.
- [ ] I also made few styles fixes (annoying scrollbar when playing for instance).

I'll create another PR if you don't mind!

mawoka-myblock commented:
Okay, so I got it installed and here are some things I noticed:

- The music upload does not work
	- Why create a new endpoint for the music upload? Why not resuse the video one?
	- After the upload, the popup closes but nothing happens on in the editor
- YouTube URL input definitely needs some cosmetic work
- YouTube doesn't autoplay on the play page
- YouTube video isn't centered
- The "Old School" mode doesn't have the YouTube video on the player's screens

I could fix them myself, though some need an explanation on your side ;)

mawoka-myblock commented:
The "view" page also seems to be missing the Youtube player, just like the "Practice" mode

Titiftw commented:
Hi @mawoka-myblock, let me answer your questions:

- The music upload does not work
  - Why create a new endpoint for the music upload? Why not resuse the video one? => Because I wanted to have the filename stored; otherwise it is complicated to find back the song in the library - which I didn't find supported by the video upload endpoint.
  - After the upload, the popup closes but nothing happens on in the editor => You're right, I typically then used the Library to add the music onto the slide. I'll have a look.
- YouTube URL input definitely needs some cosmetic work => I'll propose something but styling is not my cup of tea.
- YouTube doesn't autoplay on the play page => I'll provide a fix.
- YouTube video isn't centered => Same.
- The "Old School" mode doesn't have the YouTube video on the player's screens => Same.
- The "view" page also seems to be missing the Youtube player, just like the "Practice" mode => Same.
